Code/GraphMol/FileParsers/MolFileWriter.cpp

```diff
@@ -1053,13 +1059,14 @@ void appendEnhancedStereoGroups(std::string &res, const RWMol &tmol) {
 std::string outputMolToMolBlock(const RWMol &tmol, int confId,
                                 bool forceV3000) {
   std::string res;
-
   bool isV3000;
-  unsigned int nAtoms, nBonds, nLists, chiralFlag, nsText, nRxnComponents;
+  unsigned int nAtoms, nBonds, nLists, nSGroups, chiralFlag, nsText,
+      nRxnComponents;
   unsigned int nReactants, nProducts, nIntermediates;
   nAtoms = tmol.getNumAtoms();
   nBonds = tmol.getNumBonds();
   nLists = 0;
+  nSGroups = getNumSGroups(tmol);
```

Comment:
Having more than 999 SGroups is only a problem for V2000 SD files, in a similar way to having more than 999 atoms or bonds.

A few lines below the one you highlighted there is a check implemented for atoms and bonds, which forces using V3000 format in case of having more atoms or bonds than that. I'll add a check for SGroups too (although I though I had already done that).